### PR TITLE
UPSTREAM: 37986: Add `clusterid`, an optional parameter to storageclass.

### PR DESCRIPTION
**Pull request by humblec**
Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

### CODE REVIEW

**Comment by humblec** [comment]
@childsb @rootfs @jsafrane PTAL.

**Comment by rootfs** [comment]
@childsb @screeley44

**Comment by ncdc** [comment]
Do you need this in 1.5 or can it wait until 1.6?

**Comment by eparis** [comment]
he wants it in 1.5

File: vendor/k8s.io/kubernetes/pkg/volume/glusterfs/glusterfs.go [hunk 1]
```diff
@@ -676,7 +696,10 @@ func (p *glusterfsVolumeProvisioner) CreateVolume(gid int) (r *api.GlusterfsVolu
 		glog.Errorf("glusterfs: failed to create glusterfs rest client")
 		return nil, 0, fmt.Errorf("failed to create glusterfs REST client, REST server authentication failed")
 	}
-	clusterIds := dstrings.Split(p.clusterId, ",")
+	if p.provisioningConfig.clusterId != "" {
+		clusterIds = dstrings.Split(p.clusterId, ",")
+		glog.V(4).Infof("glusterfs: provided clusterids: %v", clusterIds)
+	}
 	gid64 := int64(gid)
 	volumeReq := &gapi.VolumeCreateRequest{Size: sz, Clusters: clusterIds, Gid: gid64, Durability: gapi.VolumeDurabilityInfo{Type: durabilityType, Replicate: gapi.ReplicaDurability{Replica: replicaCount}}}
```
Comment (screeley44) [inline vendor/k8s.io/kubernetes/pkg/volume/glusterfs/glusterfs.go]: do we want to implement the recent bug fix/change in Kube https://github.com/kubernetes/kubernetes/pull/39844

so we can use the built volumeType from parseClassParameters?
```
volumeReq := &gapi.VolumeCreateRequest{Size: sz, Clusters: clusterIds, Gid: gid64, Durability: p.volumeType}
```

Comment (humblec) [inline vendor/k8s.io/kubernetes/pkg/volume/glusterfs/glusterfs.go]: @screeley44 no. I am still not convinced on that fix. We will have a discussion on this and will act accordingly.

**Comment by humblec** [comment]
@ncdc yes, as @eparis said, we would like to have it in 3.5.

File: vendor/k8s.io/kubernetes/pkg/volume/glusterfs/glusterfs.go (hunk 1 above)
Comment (screeley44) [inline vendor/k8s.io/kubernetes/pkg/volume/glusterfs/glusterfs.go]: @humblec - fair enough, but I'm wondering why are we building the volumeType in the parseClassParameters if we will not use that config object?

**Comment by rootfs** [comment]
LGTM 

@screeley44 @humblec  if upstream 39844 is needed, can you open follow up pr?

**Comment by wongma7** [comment]
lgtm

**Comment by childsb** [comment]
[test]

**Comment by humblec** [comment]
@rootfs sure.. I am discussing with @screeley44 . Will backport it accordingly.

**Comment by rootfs** [comment]
please hold merge.

**Comment by rootfs** [comment]
lgtm

**Comment by childsb** [comment]
talked with humble and hchen, both are g2g on this now.  PR looks good and is merged upstream.  thanks guys.

[merge]

**Pull request merged**